File: src/main/python/AoC2024_08.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 8
#

import itertools
import sys
from collections import defaultdict
from functools import reduce
from operator import ior
from typing import Callable
from typing import Iterator

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.geometry import Position
from aoc.geometry import Vector

Antenna = Position
AntennaPair = tuple[Antenna, Antenna]
Input = tuple[int, int, list[set[Antenna]]]
Output1 = int
Output2 = int


TEST = """\
............
........0...
.....0......
.......0....
....0.......
......A.....
............
............
........A...
.........A..
............
............
"""


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        lines = list(input_data)
        h, w = len(lines), len(lines[0])
        antennae = defaultdict[str, set[Antenna]](set)
        for r, c in itertools.product(range(h), range(w)):
            if (freq := lines[r][c]) != ".":
                antennae[freq].add(Antenna(c, h - r - 1))
        return h, w, list(antennae.values())

    def get_antinodes(
        self, pair: AntennaPair, h: int, w: int, max_count: int = sys.maxsize
    ) -> Iterator[Antenna]:
        vec = Vector.of(pair[0].x - pair[1].x, pair[0].y - pair[1].y)
        for pos, d in itertools.product(pair, {-1, 1}):
            for a in range(1, max_count + 1):
                antinode = pos.translate(vec, d * a)
                if 0 <= antinode.x < w and 0 <= antinode.y < h:
                    yield antinode
                else:
                    break

    def solve(
        self,
        antennae: list[set[Antenna]],
        collect_antinodes: Callable[[AntennaPair], set[Antenna]],
    ) -> int:
        pairs_with_same_frequency = (
            pair
            for same_frequency in antennae
            for pair in itertools.combinations(same_frequency, 2)
        )
        return len(
            reduce(ior, map(collect_antinodes, pairs_with_same_frequency))
        )

    def part_1(self, input: Input) -> Output1:
        h, w, antennae = input

        return self.solve(
            antennae,
            lambda pair: set(self.get_antinodes(pair, h, w, max_count=1))
            - {
                pair[0],
                pair[1],
            },
        )

    def part_2(self, input: Input) -> Output2:
        h, w, antennae = input

        return self.solve(
            antennae, lambda pair: set(self.get_antinodes(pair, h, w))
        )

    @aoc_samples(
        (
            ("part_1", TEST, 14),
            ("part_2", TEST, 34),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 8)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_08.java
import static com.github.pareronia.aoc.IterTools.combinations;
import static com.github.pareronia.aoc.IterTools.product;
import static com.github.pareronia.aoc.SetUtils.difference;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.function.Function;
import java.util.stream.Stream;

import com.github.pareronia.aoc.SetUtils;
import com.github.pareronia.aoc.geometry.Position;
import com.github.pareronia.aoc.geometry.Vector;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_08
        extends SolutionBase<AoC2024_08.Input, Integer, Integer> {
    
    private AoC2024_08(final boolean debug) {
        super(debug);
    }
    
    public static AoC2024_08 create() {
        return new AoC2024_08(false);
    }
    
    public static AoC2024_08 createDebug() {
        return new AoC2024_08(true);
    }
    
    @Override
    protected Input parseInput(final List<String> inputs) {
        final int h = inputs.size();
        final int w = inputs.get(0).length();
        final Map<Character, List<Position>> antennae = new HashMap<>();
        for (int r = 0; r < h; r++) {
            for (int c = 0; c < w; c++) {
                final char ch = inputs.get(r).charAt(c);
                if (ch != '.') {
                    antennae.computeIfAbsent(ch, k -> new ArrayList<>())
                        .add(Position.of(c, h - r - 1));
                }
            }
        }
        return new Input(w, h, antennae.values().stream().toList());
    }
    
    private int solve(
            final int h,
            final int w,
            final List<List<Position>> antennae,
            final Mode mode
    ) {
        final Function<List<Position>, Stream<AntennaPair>> antennaPairs =
            sameFrequency ->
                combinations(sameFrequency.size(), 2).stream()
                    .map(comb_idx -> new AntennaPair(
                            sameFrequency.get(comb_idx[0]),
                            sameFrequency.get(comb_idx[1])));
        return antennae.stream()
                .flatMap(antennaPairs)
                .map(pair -> mode.collectAntinodes(h, w, pair))
                .reduce(SetUtils::union)
                .map(Set::size).orElseThrow();
    }
    
    @Override
    public Integer solvePart1(final Input input) {
        return solve(input.h, input.w, input.antennae, Mode.MODE_1);
    }
    
    @Override
    public Integer solvePart2(final Input input) {
        return solve(input.h, input.w, input.antennae, Mode.MODE_2);
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "14"),
        @Sample(method = "part2", input = TEST, expected = "34"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2024_08.create().run();
    }

    enum Mode {
        MODE_1, MODE_2;

        private Set<Position> getAntinodes(
                final AntennaPair pair,
                final int h,
                final int w,
                final int maxCount
        ) {
            final Vector vec = Vector.of(
                    pair.first.getX() - pair.second.getX(),
                    pair.first.getY() - pair.second.getY());
            final Set<Position> antinodes = new HashSet<>();
            product(pair, Set.of(1, -1)).stream().forEach(pp -> {
                for (int a = 1; a <= maxCount; a++) {
                    final Position antinode
                            = pp.first().translate(vec, pp.second() * a);
                    if (0 <= antinode.getX() && antinode.getX() < w
                            && 0 <= antinode.getY() && antinode.getY() < h) {
                        antinodes.add(antinode);
                    } else {
                        break;
                    }
                }
            });
            return antinodes;
        }

        public Set<Position> collectAntinodes(
                final int h, final int w, final AntennaPair pair
        ) {
            return switch (this) {
                case MODE_1 -> difference(
                                    getAntinodes(pair, h, w, 1),
                                    Set.of(pair.first, pair.second));
                case MODE_2 -> getAntinodes(pair, h, w, Integer.MAX_VALUE);
            };
        }
    }

    private static final String TEST = """
            ............
            ........0...
            .....0......
            .......0....
            ....0.......
            ......A.....
            ............
            ............
            ........A...
            .........A..
            ............
            ............
            """;

    record Input(int w, int h, List<List<Position>> antennae) {}

    record AntennaPair(Position first, Position second) implements Iterable<Position> {

        @Override
        public Iterator<Position> iterator() {
            return Set.of(first, second).iterator();
        }
    }
}